File: data/calibration_3d.py
import numpy as np
import cv2


source = np.empty([1,3],np.float32)#eros(shape, np.float32)
target = np.empty([1,3],np.float32)

# [x, y, z]
"""
source[0][0] = [857, 120, 854]
source[0][1] = [254, 120, 855]
source[0][2] = [256, 120, 255]
source[0][3] = [858, 120, 255]
target = source * 10
"""
path = "/home/ir/perception_ws/src/vicon_with_camera/data/"
img_p = path + "points%d.txt"
vic_p = path + "vicon%d.txt"
for i in range(0,10):
    si = img_p % (i)
    sv = vic_p % (i)
#    target = []
#    source = []
    with open(si) as fi:
        data = fi.readlines()
        for d in data:
            w = d.split()
            s = np.array([],np.float32)
            for wi in w:
                s = np.insert(s,len(s),float(wi))
            s = np.insert(s,len(s),1)
            target = np.vstack((target,np.atleast_2d(s)))
    with open(sv) as gi:
        data = gi.readlines()
        for d in data:
            w = d.split()
            s = np.array([],np.float32)
            for wi in w:
                s = np.insert(s,len(s),float(wi))
            source = np.vstack((source,np.atleast_2d(s)))
target = target[1:target.shape[0],:]            
source = source[1:source.shape[0],:]            

retval, M, inliers = cv2.estimateAffine3D(source, target)

for i in range (0,source.shape[0]):
    d = source[i,:]
    d = np. insert(d,len(d),1)
#    print np.mat(M)
#    print np.mat(d)
#    print '\n'
    m = np.mat(M) * np.mat(d).T
    print m.T
    print target[i,:]
    print '\n'


#print M

File: data/calibration_3d.cpp
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/calib3d/calib3d.hpp>

#include <string>
#include <iostream>
#include <fstream>

using namespace std;
using namespace cv;


int main(int argc, char* argv[]){
    int numIm = atoi(argv[1]);
    string path = "/home/ir/perception_ws/src/vicon_with_camera/data/";
    string imgP = path + "points%d.txt";
    string viconP = path + "vicon%d.txt";
    string calibP = path + "calibrated.txt";
    int nM = 4;
    vector<Point3f> calibWorldPoints;// = Mat::zeros(nM*numIm,3,CV_32F);
    vector<Point3f> calibImagePoints;// = Mat::zeros(nM*numIm,3,CV_32F);
    for(int i=0;i<numIm;i++){
        char imgF[100] = "";
        char viconF[100] = "";
        sprintf(imgF,imgP.c_str(),i);
        sprintf(viconF,viconP.c_str(),i);
//        cout<<imgF<<endl<<viconF<<endl;
        ifstream vF, iF;
        vF.open(viconF);
        iF.open(imgF);
//        cout<<viconF<<" "<<imgF<<endl;
        for(int j=0;j<nM;j++){
            float tempx, tempy, tempz, tempxi, tempyi;
            vF>>tempx>>tempy>>tempz;
//            cout<<i<<" "<<j<<" "<<i*numIm+j<<" "<<calibWorldPoints.rows<<endl;
            Point3f a, b;
            a.x = tempx;
            a.y = tempy;
            a.z = tempz;
            calibWorldPoints.push_back(a);

            iF>>tempxi>>tempyi;
            b.x = tempxi;
            b.y = tempyi;
            b.z = 1.0;
            calibImagePoints.push_back(b);
        }
        vF.close();
        iF.close();
    }
//    cout<<calibWorldPoints<<endl;
//    cout<<calibImagePoints<<endl;
    Mat cameraM = Mat::zeros(3,3,CV_32F), distC = Mat::zeros(1,5,CV_32F);
    cameraM.at<float>(0,0) = 1012.3213482211537;
    cameraM.at<float>(0,2) = 970.4085456062247;
    cameraM.at<float>(1,1) = 1019.9488892836728;
    cameraM.at<float>(1,2) = 523.8022374391905;
    cameraM.at<float>(2,2) = 1.0;

    Mat rvec, tvec;
//    solvePnPRansac(calibWorldPoints,calibImagePoints,cameraM,distC,rvec,tvec);
    Mat M = Mat::zeros(3,4,CV_32F);
    Mat inl = Mat::zeros(calibImagePoints.size(),1,CV_8U);
    estimateAffine3D(calibWorldPoints,calibImagePoints, M,inl,3,0.9);
//    solvePnP(calibWorldPoints,calibImagePoints,cameraM,distC,rvec,tvec);
    ofstream fp;
    fp.open(calibP.c_str());
    for(int i=0;i<3;i++)
        fp<<rvec.at<double>(i)<<" ";
    for(int i=0;i<3;i++)
        fp<<tvec.at<double>(i)<<" ";
    fp.close();
    return 0;
}
